### ISSUE TITLE
A wrapper to implement other tensorflow functions

### ISSUE THREAD

**Issue by bhattsamir** [greta-dev/greta#109]
There must be an easy way, I have seen in the code the op() being used.

This should save you having to add Tensor flow functions everytime the package gets updated.

**Comment by goldingn** [comment]
Yeah, I've thought about this in the past. If the function has one input and the same as the input, then it's no problem, we could do something like:

```r
tf_op <- function (tf_fun) {
  
  # make the op name
  name <- paste0("tf$", tf_fun$func_name)
  
  # return the wrapper function for greta arrays
  function (x) {
    op(name,
       x,
       tf_operation = tf_fun)
  }
  
}
```

and use it like this:

```r
x <- normal(0, 1)
square <- tf_op(tf$square)
z <- square(x)
```

The tricky bits are predicting what the dimensions of the function output will be, and handling other arguments. Unfortunately we can't automatically detect either of those things. We could let users pass that information as arguments to `tf_op`, but then the code would be even more unwieldy than constructing a function using `op`. So I'm not convinced this is a good idea.

---

One of my major design aims with greta is to avoid end users having to deal with gory coding details, but enable other developers to implement them easily in packages that extend greta. That way someone who understands the specific problem well can implement a feature once and well, and everyone else can use it. I really want to avoid end users having to recode stuff like [this](https://github.com/stan-dev/example-models/blob/master/misc/multivariate-probit/probit-multi-good.stan) (when someone could that as a `multivariate_probit()` function) or have to understand tricky things like [jacobian adjustments](https://groups.google.com/forum/#!topic/stan-users/TR-8NetrWyg). It creates a two-tier system where the software is simple for new users to a point, but then requires masses of technical knowledge to progress. I'd rather users were focussed on the models than the underlying code.

The longhand version of the `tf_op` call above would be:
```r
square <- function (x)
  op("square", x, tf$square)
```
and advanced users could always do that themselves, though I'd rather they then packaged up and shared the code for others to use. So I think I'd rather the useful functions were explicitly ported in greta (if they are common statsy things) or an extension package (for neural networky things).

**Comment by bhattsamir** [comment]
Makes perfect sense thanks
